File: part3/exercise2.go
package main

import "fmt"

// fibonacci is a function that returns
// a function that returns an int.
//使用一个闭包函数来实现最经典的斐波那契函数
func fibonacci() func() int {
	sum1 := 0
	sum2 := 1
	return func() int {
		res := sum1 + sum2
		sum1 = sum2
		sum2 = res
		return res
	}
}

func main() {
	f := fibonacci()
	for i := 0; i < 10; i++ {
		fmt.Println(f())
	}
}

File: part3/exercise2.js
const fibonacci = () => {
    let sum1 = 0
    let sum2 = 1
    return () => {
        let res = sum1 + sum2
        sum1 = sum2
        sum2 = res
        return res
    }
}
let f = fibonacci()
for (let i = 1; i <= 10; i++) {
    console.log(f())
}

//es6版fibonacci function